### PR TITLE
Make reaper service a gradle build service

### PR DESCRIPTION
**Pull request by breskeby**
- Allows creating the service lazy only when required
- Removes the need for build finish hook usage as Gradle takes care of
cleaning up
- This helps us getting closer to be gradle configuration cache compliant

### CODE REVIEW

File: buildSrc/src/main/java/org/elasticsearch/gradle/ReaperService.java [hunk 1]
```diff
@@ -24,24 +28,20 @@
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
-public class ReaperService {
+public abstract class ReaperService implements BuildService<ReaperService.Params>, AutoCloseable {
+
+    interface Params extends BuildServiceParameters {
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/ReaperService.java:33]: nit: can we throw this at the bottom of the class as it typical with other inner-classes?

File: buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java [hunk 2]
```diff
@@ -27,14 +28,21 @@ public void apply(Project project) {
...
-        ReaperService service = project.getExtensions()
-            .create("reaper", ReaperService.class, project, project.getBuildDir().toPath(), inputDir);
+            .resolve("build-" + ProcessHandle.current().pid())
+            .toFile();
 
-        project.getGradle().buildFinished(result -> service.shutdown());
+        var reaperServiceProvider = project.getGradle().getSharedServices().registerIfAbsent("reaper", ReaperService.class, spec -> {
+            // Provide some parameters
+            spec.getParameters().getInputDir().set(inputDir);
+            spec.getParameters().getBuildDir().set(project.getBuildDir());
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java:41]: Would the more correct thing be to use `project.getLayout().getBuildDirectory()` here so it's providers all the way through?

File: buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java [hunk 3]
```diff
@@ -27,14 +28,21 @@ public void apply(Project project) {
...
+            spec.getParameters().getInternal().set(BuildParams.isInternal());
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java:42]: I'm not sure there's any benefit to using a `Property` here. I think with managed types we can still use raw `boolean` which would just simplify stuff in the reaper service itself.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java [hunk 4]
```diff
@@ -66,10 +67,10 @@ public void apply(Project project) {
 
         project.getRootProject().getPluginManager().apply(ReaperPlugin.class);
 
-        ReaperService reaper = project.getRootProject().getExtensions().getByType(ReaperService.class);
+        var reaperServiceProvider = project.getRootProject().getExtensions().getByType(ReaperExtension.class).getReaperService();
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java:70]: I have mixed feelings on using `var` mixed with explicit types.

File: buildSrc/src/main/java/org/elasticsearch/gradle/ReaperExtension.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gradle;
+
+import org.gradle.api.provider.Provider;
+
+public class ReaperExtension {
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/ReaperExtension.java:13]: This is an interesting pattern. For the docker stuff I actually added a utility for looking up global services services by type:
https://github.com/elastic/elasticsearch/blob/bae65dde3ff788673fb9d3607486ba1bdce07d3d/buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java#L54-L57

I'm wondering which we should use? I don't have strong feelings either way but we should probably make them both consistent.

**Review by mark-vieira — APPROVED** [review]
Couple of comments but otherwise LGTM

File: buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java (hunk 2 above)
Comment (breskeby) [reply buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java:41]: yes

File: buildSrc/src/main/java/org/elasticsearch/gradle/ReaperExtension.java (hunk 5 above)
Comment (breskeby) [reply buildSrc/src/main/java/org/elasticsearch/gradle/ReaperExtension.java:13]: Interesting. Haven't' seen that. I'm all in for consistency here.  I'm moving this PR to use your pattern. My reasoning:

- Having an extension just to carry a property around seems a bit wrong to me. Your approach is more direct without polluting the area of surface within the DSL (Meaning we should avoid where possible to introduce extensions where the user cannot or is not intended to do anything with them)
 
Though there's some more risk to hit internal api changes with the dockerSupport example I guess. But let's see. We can always move back if we hit a barrier.

File: buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java (hunk 4 above)
Comment (breskeby) [reply buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java:70]: fixed. mixing them is probably bad indeed. I'll move forward using them more though as it makes things soo convenient and more readable

File: buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java (hunk 2 above)
Comment (breskeby) [reply buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java]: some clarification. It's better to use an injected ProjectLayout instance here (which I changed here to do so). This way we don't rely on the mutable `Project` object at all.

File: buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java (hunk 3 above)
Comment (breskeby) [reply buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java]: done

File: buildSrc/src/main/java/org/elasticsearch/gradle/ReaperExtension.java (hunk 5 above)
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/ReaperExtension.java]: > Though there's some more risk to hit internal api changes with the dockerSupport example I guess. But let's see. We can always move back if we hit a barrier.

Indeed. The APIs seems to be missing a convenient way to subsequently retrieve these global services. Ideally the would work with `@Inject`. We might want to chat with the build tool folks about what is the idiomatic way to use these things.

**Pull request merged**